Performing tasks

Naman-sopho opened the pull request:
- Oreon "at work" animation now runs when the Oreon is in the assigned area and working on a task. The Oreon also moves around randomly in the assigned area.(currently just for `plant` task)
- The `OreonCrop` blocks are now placed after the end of a `plant` task.
Edit:
Also, added the build task implementation and a building is now placed in the selected area.

Outstanding before merge : 
- [x] The Oreon crops placed in the world are not following the growth stages as they should after attaching a `ChangingBlocksComponent`

Naman-sopho committed:
Initial implementation for performing tasks

Naman-sopho committed:
Merge branch 'Oreon-health' into Performing-tasks

Naman-sopho committed:
Implement basic plant task behavior.

Add taskCompletionTime field to TaskComponent and logic for its calculation in TaskManagementSystem.
Implement new CheckTaskStatus node to check if the taskCompletionTime has passed.

Naman-sopho committed:
Oreons now move in random directons in the selected area while performing a task

Naman-sopho committed:
Add OreonCrop block to the selected area when a plant task finishes.

Add OreonCrop changing block.

Naman-sopho committed:
Add build behavior tree.

Refactor PerformTaskNode and implement separate methods for each task type.
Add dependency on DynamicCities.

Naman-sopho committed:
Remove dependency on DynamicCities.

A builiding now placed in the world after a Build task is completed.

Naman-sopho committed:
Implement building construction using StructureTemplates.

Add inn and shop templates.

Naman-sopho committed:
Debug buildings being constructed outside the selected area.

Add new food block.

mkienenb commented on src/main/java/org/terasology/healthSystem/OreonHealthSystem.java:
```diff
@@ -0,0 +1,104 @@
+        OreonAttributeComponent oreonAttributeComponent = oreon.getComponent(OreonAttributeComponent.class);
+
+        switch (cause) {
+            case "hunger" :
```
I remember that there was no way to set these up as an enum type.
However, you should at least be using "public static final String" constants to reference these states.

mkienenb commented on src/main/java/org/terasology/spawning/SpawningAuthoritySystem.java:
```diff
@@ -78,6 +81,12 @@ public void oreonSpawn(OreonSpawnEvent event, EntityRef player) {
+            CreatureNameComponent oreonNameComponent = newOreon.getComponent(CreatureNameComponent.class);
+            NameTagComponent nameTagComponent = new NameTagComponent();
+            nameTagComponent.text = oreonNameComponent.firstName;
+            nameTagComponent.yOffset = 0.5f;
```
0.5f should also be a constant.

mkienenb commented on src/main/java/org/terasology/taskSystem/actions/PerformTaskNode.java:
```diff
@@ -122,4 +205,354 @@ private void changeOreonAttributes(Actor oreon, TaskComponent taskComponent) {
+        }
+    }
+
+    private void placeCrops(Region3i selectedRegion, String cropToPlace) {
```
Move task to a separate class for crop tasks.

mkienenb commented on src/main/java/org/terasology/taskSystem/actions/PerformTaskNode.java:
```diff
@@ -122,4 +205,354 @@ private void changeOreonAttributes(Actor oreon, TaskComponent taskComponent) {
+        }
+    }
+
+    private void constructBuilding(Region3i selectedRegion, BuildingType buildingType) {
```
Move class to a separate class for constructing buildings task.
Move building-specific code to a separate class.  Possibly the raster class.

mkienenb commented on src/main/java/org/terasology/taskSystem/actions/PerformTaskNode.java:
```diff
@@ -122,4 +205,354 @@ private void changeOreonAttributes(Actor oreon, TaskComponent taskComponent) {
+        }
+    }
+
+    private void constructBuildingUsingStructureTemplates(Region3i selectedRegion, BuildingType buildingType) {
```
move to separate class

mkienenb commented on src/main/java/org/terasology/taskSystem/actions/PerformTaskNode.java:
```diff
@@ -122,4 +205,354 @@ private void changeOreonAttributes(Actor oreon, TaskComponent taskComponent) {
+        int minZ = selectedRegion.minZ();
+        int maxZ = selectedRegion.maxZ();
+
+        EntityRef template = structureTemplateProvider.getRandomTemplateOfType("MasterOfOreon:diner");
```
"MasterOfOreon:diner" should be a constant which is passed into this method.

mkienenb commented on src/main/java/org/terasology/taskSystem/actions/PerformTaskNode.java:
```diff
@@ -122,4 +205,354 @@ private void changeOreonAttributes(Actor oreon, TaskComponent taskComponent) {
+
+        EntityRef template = structureTemplateProvider.getRandomTemplateOfType("MasterOfOreon:diner");
+
+        logger.info("Placing Builiding : " + template.getParentPrefab().getName());
```
Misspelled Building.

mkienenb commented on src/main/java/org/terasology/tooltip/TooltipClientSystem.java:
```diff
@@ -30,8 +31,9 @@
-    public void addAttributesToTooltip(GetItemTooltip event, EntityRef entity, OreonAttributeComponent oreonAttributeComponent) {
+    @ReceiveEvent(components = {OreonAttributeComponent.class, CreatureNameComponent.class})
+    public void addAttributesToTooltip(GetItemTooltip event, EntityRef entity, OreonAttributeComponent oreonAttributeComponent,  CreatureNameComponent oreonNameComponent) {
+        event.getTooltipLines().add(new TooltipLine("Name : " + oreonNameComponent.firstName + " " + oreonNameComponent.lastName));
```
These attribute names need to be localized / internationalized / translated.
I'd recommend using constants for these.  I don't recall how translation works for Terasology off the top of my head.
There's probably past non-log display areas where you should also be doing translations.

Naman-sopho committed:
Refactor PerformTaskNode to move task completion method into their own classes.

Add classes ConstructFromStructureTemplate, ConstructFromBuildingGenerator and PlantTaskCompletion.

Naman-sopho committed:
Add HealthReductionCause enum

mkienenb commented on src/main/java/org/terasology/buildings/BasicRasterTarget.java:
```diff
@@ -0,0 +1,66 @@
+        this.affectedArea = area;
+        this.blockTheme = blockTheme;
+
+        Vector3i min = new Vector3i(affectedArea.minX(), -255, affectedArea.minY());
```
It seems odd that this goes from -255 to +255.

mkienenb commented on src/main/java/org/terasology/managerInterface/ManagerInterfaceSystem.java:
```diff
@@ -79,7 +79,13 @@ public void onPlayerSpawn(OnPlayerSpawnedEvent event, EntityRef player, Inventor
+            player.addComponent(new HoldingComponent());
+        }
+
+        // Added for testing purposes
```
Maybe change this to // TODO: Added for testing purposes so it is tagged and will be removed later.

mkienenb commented on src/main/java/org/terasology/taskSystem/actions/PerformTaskNode.java:
```diff
@@ -41,7 +47,18 @@
-    EntityManager entityManager;
+    private Context context;
+
+    private PlantTaskCompletionSystem plantTaskCompletion;
```
Do these need to be implemented as systems?
I was previously suggesting that they be in separate classes, but not in separate systems.
Ie,
```
    PlantTask plantTask = new PlantTask();
    plantTask.placeCrops(selectedRegion, Constants.OREON_CROP_PREFAB);
```

namtiw replied:
I implemented these as systems so we do not have to create a new instance of the task completion classes with every task.

mkienenb replied:
There's nothing wrong with creating an instance of a class, running a method on it, and then letting it be garbage collected.   This is cheap operation.   This happens every time you do something like `String x = "first" + "last";`

Keeping the class around in memory as part of the Context has a higher eventual cost.  It will never be garbage collected.   It will be one more item that has to be searched through when pulling objects out of the data context.    There is also the possibility of multithreading issues if you add an instance variable to your class and reuse it for several oreons.   That's not currently the case, though.

Note that there is not a huge cost difference either way.

Another approach if you do want to have a single class which persists is to declare the variable without @In inside performTaskNode and create the class the first time it is used in that instance variable.    Have a getPlantTaskCompletion() method which creates a new PlantTaskCompletion object if the instance variable plantTaskCompletion is null.

```
private PlantTaskCompletion getPlantTaskCompletion() {
    if (null == plantTaskCompletion) {
        plantTaskCompletion = new PlantTaskCompletion();
    }

    return plantTaskCompletion;
}
```

My recommendation is that systems be saved for API points.    Use a system if there are system methods which would eventually be accessed by another developer.

I would also rename this to PlantingTask rather than PlantTaskCompletion.   I would put all aspects of working with a Planting task into that object.   Same with other tasks.

Finally, I would pass into PlantingTask the variables that you are currently using as instance variables (blockManager, blockEntityRegistry).   If you start writing unit tests at some point, this will make it easier to test your code as well as keep your code safe for multithreading.

```
    plantTask.placeCrops(blockManager, blockEntityRegistry, selectedRegion, Constants.OREON_CROP_PREFAB);
```

Naman-sopho committed:
Modify implementation of task completion classes
